src/main/java/com/aws/greengrass/logmanager/services/DiskSpaceManagementService.java

```diff
@@ -0,0 +1,219 @@
+import java.nio.file.WatchService;
+import java.nio.file.Watchable;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+public class DiskSpaceManagementService {
+    private static final Logger logger = LogManager.getLogger(DiskSpaceManagementService.class);
+
+    private final Object spaceManagementLock = new Object();
+    private Future<?> spaceManagementThread;
+    private final ExecutorService executorService;
+
+
+    /**
+     * Constructor.
+     *
+     * @param executorService {@link ExecutorService}
+     */
+    public DiskSpaceManagementService(ExecutorService executorService) {
+        this.executorService = executorService;
+    }
+
+
+    public void start(Map<String, ComponentLogConfiguration> componentLogConfigurations) {
+        this.scheduleSpaceManagementThread(componentLogConfigurations);
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingThrowable")
+    private void scheduleSpaceManagementThread(Map<String, ComponentLogConfiguration> componentLogConfigurations) {
+        synchronized (spaceManagementLock) {
+            if (spaceManagementThread != null) {
+                spaceManagementThread.cancel(true);
+            }
+            spaceManagementThread = this.executorService.submit(() -> {
+                try {
+                    logger.atInfo().log("Starting space management thread");
+                    startWatchServiceOnLogFilePaths(componentLogConfigurations);
+                } catch (IOException e) {
+                    //TODO: fail the deployment?
+                    logger.atError().cause(e).log("Unable to start space management thread.");
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                } catch (Throwable e) {
+                    logger.atError().log("Failure in log manager space management", e);
+                    scheduleSpaceManagementThread(componentLogConfigurations); // restart space management
+                }
+            });
+        }
+    }
+
+    /**
+     * Starts a Watch Service on all the log directories where customer has specified disk space limit.
+     * If any file is modified and created in any of those directories, we get an event containing that information.
+     * The log manager will then verify that the size of all log files in that directory for the component does not
+     * exceed the disk space limit set by the customer. If it does, the log manager will delete the log files until
+     * the limit is met.
+     *
+     * @throws IOException          if unable to initialise a new Watch Service.
+     * @throws InterruptedException if thread is shutdown
+     */
+    private void startWatchServiceOnLogFilePaths(Map<String, ComponentLogConfiguration> componentLogConfigurations)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.